File: DL-ATM41/DL-ATM41.erl

% https://www.decentlab.com/products/eleven-parameter-weather-station-for-lorawan

-module(decentlab_decoder).
-define(PROTOCOL_VERSION, 2).
-export([decode/1, start/0]).

-define(SENSOR_DEFS,
  [
    #{
      length => 17,
      values => [
        #{
          name => <<"Solar radiation">>,
          convert => fun(X) -> lists:nth(0 + 1, X) - 32768 end,
          unit => <<"W⋅m⁻²"/utf8>>
        },
        #{
          name => <<"Precipitation">>,
          convert => fun(X) -> (lists:nth(1 + 1, X) - 32768) / 1000 end,
          unit => <<"mm"/utf8>>
        },
        #{
          name => <<"Lightning strike count">>,
          convert => fun(X) -> lists:nth(2 + 1, X) - 32768 end
        },
        #{
          name => <<"Lightning average distance">>,
          convert => fun(X) -> lists:nth(3 + 1, X) - 32768 end,
          unit => <<"km"/utf8>>
        },
        #{
          name => <<"Wind speed">>,
          convert => fun(X) -> (lists:nth(4 + 1, X) - 32768) / 100 end,
          unit => <<"m⋅s⁻¹"/utf8>>
        },
        #{
          name => <<"Wind direction">>,
          convert => fun(X) -> (lists:nth(5 + 1, X) - 32768) / 10 end,
          unit => <<"°"/utf8>>
        },
        #{
          name => <<"Maximum wind speed">>,
          convert => fun(X) -> (lists:nth(6 + 1, X) - 32768) / 100 end,
          unit => <<"m⋅s⁻¹"/utf8>>
        },
        #{
          name => <<"Air temperature">>,
          convert => fun(X) -> (lists:nth(7 + 1, X) - 32768) / 10 end,
          unit => <<"°C"/utf8>>
        },
        #{
          name => <<"Vapor pressure">>,
          convert => fun(X) -> (lists:nth(8 + 1, X) - 32768) / 100 end,
          unit => <<"kPa"/utf8>>
        },
        #{
          name => <<"Atmospheric pressure">>,
          convert => fun(X) -> (lists:nth(9 + 1, X) - 32768) / 100 end,
          unit => <<"kPa"/utf8>>
        },
        #{
          name => <<"Relative humidity">>,
          convert => fun(X) -> (lists:nth(10 + 1, X) - 32768) / 10 end,
          unit => <<"%"/utf8>>
        },
        #{
          name => <<"Sensor temperature (internal)">>,
          convert => fun(X) -> (lists:nth(11 + 1, X) - 32768) / 10 end,
          unit => <<"°C"/utf8>>
        },
        #{
          name => <<"X orientation angle">>,
          convert => fun(X) -> (lists:nth(12 + 1, X) - 32768) / 10 end,
          unit => <<"°"/utf8>>
        },
        #{
          name => <<"Y orientation angle">>,
          convert => fun(X) -> (lists:nth(13 + 1, X) - 32768) / 10 end,
          unit => <<"°"/utf8>>
        },
        #{
          name => <<"Compass heading">>,
          convert => fun(X) -> lists:nth(14 + 1, X) - 32768 end,
          unit => <<"°"/utf8>>
        },
        #{
          name => <<"North wind speed">>,
          convert => fun(X) -> (lists:nth(15 + 1, X) - 32768) / 100 end,
          unit => <<"m⋅s⁻¹"/utf8>>
        },
        #{
          name => <<"East wind speed">>,
          convert => fun(X) -> (lists:nth(16 + 1, X) - 32768) / 100 end,
          unit => <<"m⋅s⁻¹"/utf8>>
        }
      ]
    },
    #{
      length => 1,
      values => [
        #{
          name => <<"Battery voltage">>,
          convert => fun(X) -> lists:nth(0 + 1, X) / 1000 end,
          unit => <<"V"/utf8>>
        }
      ]
    }
  ]
).


decode(<<ProtocolVersion, DeviceId:16,
         Flags:16/bitstring, Bytes/binary>> = Binary) when is_binary(Binary) ->
  Words = bytes_to_words(Bytes),
  maps:merge(#{<<"Protocol version">> => ProtocolVersion,
               <<"Device Id">> => DeviceId},
             sensor(Words, Flags, ?SENSOR_DEFS));
decode(HexStr) ->
  Binary = hex2bin(HexStr),
  decode(Binary).


start() ->
  io:format("~p~n", [decode("02035a0003800a8000800080008009812b8014810880b4a57c820c810980027fe88056800880040bf5")]),
  io:format("~p~n", [decode("02035a00020bf5")]).


%%%%%%%%%%%%%%%%%%%%%%%%%%%%% PRIVATE FUNCTIONS %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

sensor(_Words, _Flags, []) -> #{};
sensor(Words, <<Flags:15, 0:1>>, [_Cur | Rest]) ->
  sensor(Words, <<0:1, Flags:15>>, Rest);
sensor(Words, <<Flags:15, 1:1>>,
       [#{length := Len, values := ValueDefs} | RestDefs]) ->
  {X, RestWords} = lists:split(Len, Words),
  Values = value(ValueDefs, X),
  RestSensors = sensor(RestWords, <<0:1, Flags:15>>, RestDefs),
  maps:merge(Values, RestSensors).


value([], _X) -> #{};
value([#{convert := Convert,
         name := Name} = ValueDef | RestDefs], X) ->
  Values = value(RestDefs, X),
  Value = #{<<"value">> => Convert(X)},
  Value2 = add_unit(Value, ValueDef),
  maps:put(Name, Value2, Values);
value([_NoConvert | RestDefs], X) -> value(RestDefs, X).


add_unit(Value, #{unit := Unit}) ->
  maps:put(<<"unit">>, Unit, Value);
add_unit(Value, _NoUnit) ->
  Value.


hex2bin(HexStr) when is_binary(HexStr) -> hex2bin(binary_to_list(HexStr));
hex2bin(HexStr) -> hex2bin(HexStr, []).

hex2bin("", Bins) -> list_to_binary(lists:reverse(Bins));
hex2bin(HexStr, Bins) ->
  Bytes = string:slice(HexStr, 0, 2),
  Word = list_to_integer(Bytes, 16),
  hex2bin(string:slice(HexStr, 2), [Word | Bins]).


bytes_to_words(Bytes) -> bytes_to_words(Bytes, []).

bytes_to_words(<<>>, Words) -> lists:reverse(Words);
bytes_to_words(<<Word:16, Rest/binary>>, Words) ->
  bytes_to_words(Rest, [Word | Words]).

File: DL-ATM41/DL-ATM41.php
<?php
/* https://www.decentlab.com/products/eleven-parameter-weather-station-for-lorawan */

abstract class DecentlabDecoder
{
    const PROTOCOL_VERSION = 2;

    protected $sensors;

    public function decode($payload = '')
    {
        $bytes = hex2bin($payload);
        $parts = [];

        $parts['version'] = ord($bytes[0]);
        if ($parts['version'] != self::PROTOCOL_VERSION) {
            $parts['error'] = sprintf("protocol version %u doesn't match v2", $parts['version']);
            return $parts;
        }

        $parts['device_id'] = unpack('n', substr($bytes, 1))[1];
        $flags = unpack('n', substr($bytes, 3))[1];

        /* decode payload */
        $k = 5;
        foreach ($this->sensors as $sensor) {
            if (($flags & 1) == 1) {
                $x = [];
                /* convert data to 16-bit integer array */
                for ($j = 0; $j < $sensor['length']; $j++) {
                    array_push($x, unpack('n', substr($bytes, $k))[1]);
                    $k += 2;
                }

                /* decode sensor values */
                foreach ($sensor['values'] as $value) {
                    if ($value['convert'] != NULL) {
                        $parts[$value['name'] . '_value'] = $value['convert']($x);
                        if ($value['unit'] != NULL) {
                            $parts[$value['name'] . '_unit'] = $value['unit'];
                        }
                    }
                }
            }
            $flags >>= 1;
        }

        return $parts;
    }
}

class DL_ATM41_Decoder extends DecentlabDecoder {
    
    public function __construct()
    {
        $this->sensors = [
            [
                'length' => 17,
                'values' => [
                    [
                        'name' => 'solar_radiation',
                        'convert' => function ($x) { return $x[0] - 32768; },
                        'unit' => 'W⋅m⁻²',
                    ],
                    [
                        'name' => 'precipitation',
                        'convert' => function ($x) { return ($x[1] - 32768) / 1000; },
                        'unit' => 'mm',
                    ],
                    [
                        'name' => 'lightning_strike_count',
                        'convert' => function ($x) { return $x[2] - 32768; },
                        'unit' => NULL,
                    ],
                    [
                        'name' => 'lightning_average_distance',
                        'convert' => function ($x) { return $x[3] - 32768; },
                        'unit' => 'km',
                    ],
                    [
                        'name' => 'wind_speed',
                        'convert' => function ($x) { return ($x[4] - 32768) / 100; },
                        'unit' => 'm⋅s⁻¹',
                    ],
                    [
                        'name' => 'wind_direction',
                        'convert' => function ($x) { return ($x[5] - 32768) / 10; },
                        'unit' => '°',
                    ],
                    [
                        'name' => 'maximum_wind_speed',
                        'convert' => function ($x) { return ($x[6] - 32768) / 100; },
                        'unit' => 'm⋅s⁻¹',
                    ],
                    [
                        'name' => 'air_temperature',
                        'convert' => function ($x) { return ($x[7] - 32768) / 10; },
                        'unit' => '°C',
                    ],
                    [
                        'name' => 'vapor_pressure',
                        'convert' => function ($x) { return ($x[8] - 32768) / 100; },
                        'unit' => 'kPa',
                    ],
                    [
                        'name' => 'atmospheric_pressure',
                        'convert' => function ($x) { return ($x[9] - 32768) / 100; },
                        'unit' => 'kPa',
                    ],
                    [
                        'name' => 'relative_humidity',
                        'convert' => function ($x) { return ($x[10] - 32768) / 10; },
                        'unit' => '%',
                    ],
                    [
                        'name' => 'sensor_temperature_internal',
                        'convert' => function ($x) { return ($x[11] - 32768) / 10; },
                        'unit' => '°C',
                    ],
                    [
                        'name' => 'x_orientation_angle',
                        'convert' => function ($x) { return ($x[12] - 32768) / 10; },
                        'unit' => '°',
                    ],
                    [
                        'name' => 'y_orientation_angle',
                        'convert' => function ($x) { return ($x[13] - 32768) / 10; },
                        'unit' => '°',
                    ],
                    [
                        'name' => 'compass_heading',
                        'convert' => function ($x) { return $x[14] - 32768; },
                        'unit' => '°',
                    ],
                    [
                        'name' => 'north_wind_speed',
                        'convert' => function ($x) { return ($x[15] - 32768) / 100; },
                        'unit' => 'm⋅s⁻¹',
                    ],
                    [
                        'name' => 'east_wind_speed',
                        'convert' => function ($x) { return ($x[16] - 32768) / 100; },
                        'unit' => 'm⋅s⁻¹',
                    ],
                ],
            ],
            [
                'length' => 1,
                'values' => [
                    [
                        'name' => 'battery_voltage',
                        'convert' => function ($x) { return $x[0] / 1000; },
                        'unit' => 'V',
                    ],
                ],
            ],
        ];
    }
}


$decoder = new DL_ATM41_Decoder();
$payloads = [
    '02035a0003800a8000800080008009812b8014810880b4a57c820c810980027fe88056800880040bf5',
    '02035a00020bf5',
];

foreach($payloads as $payload) {
    var_dump($decoder->decode($payload));
}
